Sources/Plasma/FeatureLib/pfGLPipeline/plGLDevice.cpp

```diff
@@ -40,29 +40,891 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+
+    if (target != nullptr) {
+        ref = static_cast<plGLRenderTargetRef*>(target->GetDeviceRef());
+
+        if (ref == nullptr || ref->IsDirty())
+            ref = static_cast<plGLRenderTargetRef*>(fPipeline->MakeRenderTargetRef(target));
+    }
+
+    if (ref == nullptr) {
+        /// Set to main screen
+        glBindFramebuffer(GL_FRAMEBUFFER, 0);
+
+        if (plGLVersion() >= 46)
+            glClipControl(GL_LOWER_LEFT, GL_ZERO_TO_ONE);
+    } else {
+        /// Set to this target
+        glBindFramebuffer(GL_FRAMEBUFFER, ref->fFrameBuffer);
+
+        // We need to flip the Y axis :(
+        if (plGLVersion() >= 46)
+            glClipControl(GL_UPPER_LEFT, GL_ZERO_TO_ONE);
+        // else... find a way to do this with the projection matrix?
+    }
+
+    SetViewport();
 }
 
 void plGLDevice::SetViewport()
 {
+    glViewport(fPipeline->GetViewTransform().GetViewPortLeft(),
+               fPipeline->GetViewTransform().GetViewPortTop(),
+               fPipeline->GetViewTransform().GetViewPortWidth(),
+               fPipeline->GetViewTransform().GetViewPortHeight());
+}
+
+bool plGLDevice::BeginRender()
+{
+#ifdef HS_BUILD_FOR_WIN32
+    // Best practice, apparently, is to get and release the DC every time we need it.
+    // A DC is only valid on one thread at a time.
+    fDevice = static_cast<hsWindowHndl>((HANDLE)GetDC(fWindow));
+#endif
+
+    if (fActiveThread == hsThread::ThisThreadHash()) {
+        return true;
+    }
+
+    fActiveThread = hsThread::ThisThreadHash();
+
+    // Initialize OpenGL
+    if (!InitDevice()) {
+        plStatusLog::AddLineS("pipeline.log", GetErrorString());
+        return false;
+    }
+
+#ifdef USE_EGL
+    if (fContextType == kEGL) {
+        EGLDisplay display = static_cast<EGLDisplay>(fDisplay);
+        EGLContext context = static_cast<EGLContext>(fContext);
+        EGLSurface surface = static_cast<EGLSurface>(fSurface);
+
+        if (eglMakeCurrent(display, surface, surface, context) == EGL_FALSE) {
+            fErrorMsg = "Failed to attach EGL context to surface";
+            return false;
+        }
+    } //else
+#endif
+
+    return true;
+}
+
+bool plGLDevice::EndRender()
+{
+    if (fPipeline->fCurrRenderTarget != nullptr) {
+        return true;
+    }
+
+#ifdef USE_EGL
+    if (fContextType == kEGL) {
+        if (eglSwapBuffers(static_cast<EGLDisplay>(fDisplay), static_cast<EGLSurface>(fSurface)) == EGL_FALSE) {
+            fErrorMsg = "Failed to swap buffers";
+            return false;
+        }
+        return true;
+    } else
+#endif
+
+#ifdef HS_BUILD_FOR_WIN32
+    if (fContextType == kWGL) {
+        SwapBuffers(static_cast<HDC>((HANDLE)fDevice));
+    }
+
+    ReleaseDC(fWindow, static_cast<HDC>((HANDLE)fDevice));
+    fDevice = nullptr;
+#endif
+
+    return false;
+}
+
+static uint32_t  IGetBufferFormatSize(uint8_t format)
+{
+    uint32_t  size = sizeof( float ) * 6 + sizeof( uint32_t ) * 2; // Position and normal, and two packed colors
```

Comment:
```suggestion
    constexpr uint32_t  size = sizeof(float) * 6 + sizeof(uint32_t) * 2; // Position and normal, and two packed colors
```

Reply:
This can't be `constexpr` because we add to the `size` variable and then return it